models/buffer_disk.go

```diff
@@ -0,0 +1,222 @@
+package models
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/tidwall/wal"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/metric"
+)
+
+type DiskBuffer struct {
+	BufferStats
+	sync.Mutex
+
+	file *wal.Log
+	path string
+
+	batchFirst uint64 // Index of the first metric in the batch
+	batchSize  uint64 // Number of metrics currently in the batch
+
+	// Ending point of metrics read from disk on telegraf launch.
+	// Used to know whether to discard tracking metrics.
+	originalEnd uint64
+}
+
+func NewDiskBuffer(name string, path string, stats BufferStats) (*DiskBuffer, error) {
+	filePath := path + "/" + name
+	walFile, err := wal.Open(filePath, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open wal file: %w", err)
+	}
+	return &DiskBuffer{
+		BufferStats: stats,
+		file:        walFile,
+		path:        filePath,
+	}, nil
+}
+
+func (b *DiskBuffer) Len() int {
+	b.Lock()
+	defer b.Unlock()
+	return b.length()
+}
+
+func (b *DiskBuffer) length() int {
+	// Special case for when the read index is zero, it must be empty (otherwise it would be >= 1)
+	if b.readIndex() == 0 {
+		return 0
+	}
+	return int(b.writeIndex() - b.readIndex())
+}
+
+// readIndex is the first index to start reading metrics from, or the head of the buffer
+func (b *DiskBuffer) readIndex() uint64 {
+	index, err := b.file.FirstIndex()
+	if err != nil {
+		panic(err) // can only occur with a corrupt wal file
+	}
+	return index
+}
+
+// writeIndex is the first index to start writing metrics to, or the tail of the buffer
+func (b *DiskBuffer) writeIndex() uint64 {
+	index, err := b.file.LastIndex()
+	if err != nil {
+		panic(err) // can only occur with a corrupt wal file
+	}
+	return index + 1
+}
+
+func (b *DiskBuffer) Add(metrics ...telegraf.Metric) int {
+	b.Lock()
+	defer b.Unlock()
+
+	dropped := 0
+	for _, m := range metrics {
+		if !b.addSingle(m) {
+			dropped++
+		}
+	}
+	b.BufferSize.Set(int64(b.length()))
+	return dropped
+	// todo implement batched writes
+}
+
+func (b *DiskBuffer) addSingle(m telegraf.Metric) bool {
+	data, err := metric.ToBytes(m)
+	if err != nil {
+		panic(err)
+	}
+	err = b.file.Write(b.writeIndex(), data)
+	if err == nil {
+		b.metricAdded()
+		return true
+	}
+	return false
+}
+
+//nolint:unused // to be implemented in the future
```

Comment:
What is this addBatch eventually going to use for?

Reply:
ok - is it worth keeping this function around then? can it get added to Batch already?

Reply:
I think we can move to exclusively this, I was considering using `addSingle` for single metric call to `Add()`, and `addBatch` for multiple metric call to avoid the extra allocation but that may be unnecessary code complication for such a minor difference

Reply:
agreed - let's keep it simple. If you aren't ready with addBatch then we can remove it and avoid unused code for now, but I'd rather just go right to that if possible.

Reply:
For now I think we need to stay with addSingle so that we can accurately report how many metrics were dropped; I'll look into an upstream change to see if we can make the batch interface work in the future